src/inmanta/ast/statements/generator.py

```diff
@@ -243,18 +248,37 @@ def __init__(self, condition: ExpressionStatement, if_branch: BasicBlock, else_b
         self.condition: ExpressionStatement = condition
         self.if_branch: BasicBlock = if_branch
         self.else_branch: BasicBlock = else_branch
+        self._if_promises: List[ConditionalPromiseBlock] = []
+        self._else_promises: List[ConditionalPromiseBlock] = []
         self.anchors.extend(condition.get_anchors())
         self.anchors.extend(if_branch.get_anchors())
         self.anchors.extend(else_branch.get_anchors())
 
     def __repr__(self) -> str:
         return "If"
 
+    # TODO: document this behavior somewhere
+    # TODO: add rationale somewhere: ? circular logic always nondeterministic in terms of model (deterministic only due to compilation order, which is out of scope of the model) ?
     def normalize(self) -> None:
         self.condition.normalize()
         self.if_branch.normalize()
         self.else_branch.normalize()
 
+    def emit_progression_promises(
+        self, resolver: Resolver, queue: QueueScheduler, *, in_scope: FrozenSet[str], root: bool = False
+    ) -> Sequence[ConditionalPromiseABC]:
+        if_block: ConditionalPromiseBlock
+        else_block: ConditionalPromiseBlock
+        if_block, else_block = (
```

Comment:
would there be any benefit in contracting cognitional promisses that exist on both side into a non-conditional? I guess not?

Reply:
I think it would be a negligible improvement but I find it hard to say for sure.